lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,133 @@
+
+import { ObservableQueue } from './ObservableQueue'
+import type { QueueDiscoverer } from './queueDiscoverers'
+
+export type Metrics = {
+  completedGauge: prometheus.Gauge<never>
+  activeGauge: prometheus.Gauge<never>
+  delayedGauge: prometheus.Gauge<never>
+  failedGauge: prometheus.Gauge<never>
+  waitingGauge: prometheus.Gauge<never>
+  completedDuration: prometheus.Histogram<never>
+  processedDuration: prometheus.Histogram<never>
+}
+
+export type MetricCollectorOptions = {
+  redisClient: Redis
+  bullMqPrefix: string
+  metricsPrefix: string
+  queueDiscoverer: QueueDiscoverer
+  excludedQueues: string[]
+  histogramBuckets: number[]
+}
+
+const getMetrics = (prefix: string, histogramBuckets: number[]): Metrics => ({
+  completedGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_completed`,
+    help: 'Total number of completed jobs',
+    labelNames: ['queue'],
+  }),
+  activeGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_active`,
+    help: 'Total number of active jobs (currently being processed)',
+    labelNames: ['queue'],
+  }),
+  failedGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_failed`,
+    help: 'Total number of failed jobs',
+    labelNames: ['queue'],
+  }),
+  delayedGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_delayed`,
+    help: 'Total number of jobs that will run in the future',
+    labelNames: ['queue'],
+  }),
+  waitingGauge: new prometheus.Gauge({
+    name: `${prefix}_jobs_waiting`,
+    help: 'Total number of jobs waiting to be processed',
+    labelNames: ['queue'],
+  }),
+  processedDuration: new prometheus.Histogram({
+    name: `${prefix}_jobs_processed_duration`,
+    help: 'Processing time for completed jobs (processing until completed)',
+    buckets: histogramBuckets,
+    labelNames: ['queue'],
+  }),
+  completedDuration: new prometheus.Histogram({
+    name: `${prefix}_jobs_completed_duration`,
+    help: 'Completion time for jobs (created until completed)',
+    buckets: histogramBuckets,
+    labelNames: ['queue'],
+  }),
+})
+
+export class MetricsCollector {
+  private readonly redis: Redis
+  private readonly metrics: Metrics
+  private observedQueues: ObservableQueue[] | undefined
+
+  constructor(
+    private readonly options: MetricCollectorOptions,
+    private readonly registry: prometheus.Registry,
+    private readonly logger: FastifyBaseLogger,
+  ) {
+    this.redis = options.redisClient
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  /**
+   * Updates metrics for all discovered queues
+   */
+  async collect() {
+    if (!this.observedQueues) {
+      this.observedQueues = (await this.options.queueDiscoverer.discoverQueues())
+        .filter((name) => !this.options.excludedQueues.includes(name))
+        .map((name) => new ObservableQueue(name, this.redis, this.metrics, this.logger))
+    }
+
+    await PromisePool.for(this.observedQueues).process(async (queue) => {
+      await queue.collect()
```

Comment:
async/await seems redundant here, you can just return the promise you get from the collect method

Reply:
yeah, but then this place won't show up in the stacktrace and I think that await will be zero-cost here: https://v8.dev/blog/fast-async

Reply:
not zero, it's still one extra micro-tick and a throwaway promise.

I don't expect any useful stacktraces in this part of the code, tbh.

no strong opinion on this, just a nit.